test/xpu_api/asynch.pass.cpp

```diff
@@ -0,0 +1,142 @@
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "oneapi/dpl/execution"
+#include "oneapi/dpl/async"
+#include "oneapi/dpl/iterator"
+
+#include "support/pstl_test_config.h"
+
+#include <iostream>
+#include <iomanip>
+
+#if TEST_SYCL_PRESENT
+#    include <CL/sycl.hpp>
+#endif
+
+template <typename _T1, typename _T2>
+void
+ASSERT_EQUAL(_T1&& X, _T2&& Y)
+{
+    if (X != Y)
+        std::cout << "CHECK CORRECTNESS (ASYNC): fail (" << X << "," << Y << ")" << std::endl;
+}
+
+#if TEST_DPCPP_BACKEND_PRESENT
+void
+test_with_buffers()
+{
+    const int n = 100;
+    {
+        sycl::queue q;
+        sycl::buffer<int> x{n};
+        sycl::buffer<int> y{n};
+
+        auto my_policy = oneapi::dpl::execution::make_device_policy(q);
+        auto res_1a = oneapi::dpl::experimental::copy_async(my_policy, oneapi::dpl::counting_iterator<int>(0),
+                                                            oneapi::dpl::counting_iterator<int>(n),
+                                                            oneapi::dpl::begin(x)); // x = [0..n]
+        auto res_1b = oneapi::dpl::experimental::fill_async(my_policy, oneapi::dpl::begin(y), oneapi::dpl::end(y),
+                                                            7); // y = [7..7]
+
+        auto res_2a = oneapi::dpl::experimental::for_each_async(
+            my_policy, oneapi::dpl::begin(x), oneapi::dpl::end(x), [](int& e) { ++e; }, res_1a); // x = [1..n]
+        auto res_2b = oneapi::dpl::experimental::transform_async(
+            my_policy, oneapi::dpl::begin(y), oneapi::dpl::end(y), oneapi::dpl::begin(y),
+            [](const int& e) { return e / 2; },
+            res_1b); // y = [3..3]
+
+        sycl::buffer<int> z{n}; //std::vector<int> z(n);
+        auto res_3 = oneapi::dpl::experimental::transform_async(my_policy, oneapi::dpl::begin(x), oneapi::dpl::end(x),
+                                                                oneapi::dpl::begin(y), oneapi::dpl::begin(z),
+                                                                std::plus<int>(), res_2a, res_2b); // z = [4..n+3]
+
+        auto alpha = oneapi::dpl::experimental::reduce_async(my_policy, oneapi::dpl::begin(x), oneapi::dpl::end(x), 0,
+                                                             std::plus<int>(),
+                                                             res_2a)
+                         .get(); // alpha = n*(n+1)/2
+
+        auto beta =
+            oneapi::dpl::experimental::transform_reduce_async(my_policy, oneapi::dpl::begin(z), oneapi::dpl::end(z), 0,
+                                                              std::plus<int>(), [=](int e) { return alpha * e; })
+                .get();
+
+        ASSERT_EQUAL(beta, (n * (n + 1) / 2) * ((n + 3) * (n + 4) / 2 - 6));
+    }
+}
+
+void
+test_with_usm()
+{
+    cl::sycl::queue q;
+    const int n = 13;
```

Comment:
I guess we should have non-trivial workload to test the async calls, to reproduce protentional races..
For example, in case of two async algoeithms, we should have big workload on first call and small workload on second async algo call, to cover 'synchronization' and 'dependency' errors.

Reply:
I've slightly extended the test, calling a sort after async transform reduce. Also added a `TODO` to extend the tests.